#Black Friday Sales Prediction

- A retail company “ABC Private Limited” wants to understand the customer purchase behaviour (specifically, purchase amount) against various products of different categories. They have shared purchase summary of various customers for selected high volume products from last month.
The data set also contains customer demographics (age, gender, marital status, city_type, stay_in_current_city), product details (product_id and product category) and Total purchase_amount from last month.

- Now, they want to build a model to predict the purchase amount of customer against various products which will help them to create personalized offer for customers against different products.

- https://datahack.analyticsvidhya.com/contest/black-friday/#ProblemStatement

- Description of data set is given in above link

In [4]:
import re
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

- importing all the required libraries and importing dataset from github account to saledf dataframe

In [5]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [6]:
Saledf = pd.read_csv('/content/drive/MyDrive/BlackFridayPredictionClean.csv')

In [7]:
Saledf.head()

User_ID Product_ID Gender  ...  Product_Category_2  Product_Category_3 Purchase
0  1000001  P00069042      F  ...                 8.0                16.0     8370
1  1000001  P00248942      F  ...                 6.0                14.0    15200
2  1000001  P00087842      F  ...                 8.0                16.0     1422
3  1000001  P00085442      F  ...                14.0                16.0     1057
4  1000002  P00285442      M  ...                 8.0                16.0     7969

[5 rows x 12 columns]

- Checking Dataset and No of rows and columns

In [8]:
row, col = Saledf.shape
row, col

(550068, 12)

- Information about dataset

In [9]:
Saledf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550068 entries, 0 to 550067
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   User_ID                     550068 non-null  int64  
 1   Product_ID                  550068 non-null  object 
 2   Gender                      550068 non-null  object 
 3   Age                         550068 non-null  int64  
 4   Occupation                  550068 non-null  int64  
 5   City_Category               550068 non-null  object 
 6   Stay_In_Current_City_Years  550068 non-null  int64  
 7   Marital_Status              550068 non-null  int64  
 8   Product_Category_1          550068 non-null  int64  
 9   Product_Category_2          550068 non-null  float64
 10  Product_Category_3          550068 non-null  float64
 11  Purchase                    550068 non-null  int64  
dtypes: float64(2), int64(7), object(3)
memory usage: 50.4+ MB


# Summary of above output

- Above dataset contains 12 columns
- There are 5 columns are in integer datatype, 2 columns are in float datatype and 5 columns are in object datatype.
- Above dataset have 550068 entries.

# Checking for null values in dataset

In [10]:
Saledf.isna().sum()

User_ID                       0
Product_ID                    0
Gender                        0
Age                           0
Occupation                    0
City_Category                 0
Stay_In_Current_City_Years    0
Marital_Status                0
Product_Category_1            0
Product_Category_2            0
Product_Category_3            0
Purchase                      0
dtype: int64

In [11]:
def Gender(x):
  if x=='F':return 0
  else: return 1

In [12]:
Saledf.Gender = Saledf.Gender.apply(Gender)

In [13]:
Saledf.head()

User_ID Product_ID  Gender  ...  Product_Category_2  Product_Category_3 Purchase
0  1000001  P00069042       0  ...                 8.0                16.0     8370
1  1000001  P00248942       0  ...                 6.0                14.0    15200
2  1000001  P00087842       0  ...                 8.0                16.0     1422
3  1000001  P00085442       0  ...                14.0                16.0     1057
4  1000002  P00285442       1  ...                 8.0                16.0     7969

[5 rows x 12 columns]

# converting categorical values to numbers for City category column

In [14]:
Saledf.City_Category.unique()

array(['A', 'C', 'B'], dtype=object)

In [15]:
citydf = pd.get_dummies(Saledf.City_Category, prefix='City')

In [16]:
Saledf1 = pd.concat([Saledf, citydf], axis=1)

In [17]:
Saledf1.drop(columns=['City_Category'], axis=1, inplace=True)
Saledf1.head()

User_ID Product_ID  Gender  Age  ...  Purchase  City_A  City_B  City_C
0  1000001  P00069042       0    0  ...      8370       1       0       0
1  1000001  P00248942       0    0  ...     15200       1       0       0
2  1000001  P00087842       0    0  ...      1422       1       0       0
3  1000001  P00085442       0    0  ...      1057       1       0       0
4  1000002  P00285442       1    6  ...      7969       0       0       1

[5 rows x 14 columns]

# Extract Train and test dataset

In [18]:
X= Saledf1.drop(columns=['Purchase','Product_ID'],axis=1)
y= Saledf1.Purchase.values.reshape(-1,1)

In [19]:
X.shape,y.shape

((550068, 12), (550068, 1))

In [20]:
from sklearn.model_selection import train_test_split

In [21]:
X_train,X_test,y_train,y_test = train_test_split(
    X,y,test_size=0.25, random_state=1234
)

In [22]:
X_train.shape,X_test.shape,y_train.shape,y_test.shape

((412551, 12), (137517, 12), (412551, 1), (137517, 1))

# Linear Regression Model

In [23]:
from sklearn.linear_model import LinearRegression

In [24]:
model = LinearRegression()

# Cross Validation Technique

In [25]:
from sklearn.model_selection import KFold

In [26]:
from sklearn.model_selection import cross_val_score

In [27]:
X_new = pd.concat([X_train,X_test])
X_new.head()

User_ID  Gender  Age  ...  City_A  City_B  City_C
386133  1005433       0    3  ...       0       1       0
216879  1003499       1    2  ...       1       0       0
507279  1000131       1    1  ...       0       0       1
231169  1005648       1    2  ...       0       0       1
333308  1003390       0    4  ...       0       0       1

[5 rows x 12 columns]

In [28]:
X_new.shape

(550068, 12)

In [29]:
X_new = X_new.reset_index().drop('index',axis=1)

In [30]:
X_new.head()

User_ID  Gender  Age  Occupation  ...  Product_Category_3  City_A  City_B  City_C
0  1005433       0    3          17  ...                16.0       0       1       0
1  1003499       1    2           3  ...                16.0       1       0       0
2  1000131       1    1           4  ...                16.0       0       0       1
3  1005648       1    2          17  ...                16.0       0       0       1
4  1003390       0    4          20  ...                16.0       0       0       1

[5 rows x 12 columns]

In [31]:
y_new = pd.concat([pd.DataFrame(y_train,columns=['Price']),pd.DataFrame(y_test,columns=['Price'])])
y_new.head()

Price
0   4575
1  15581
2   5188
3   9767
4  13702

In [32]:
y_new.shape

(550068, 1)

In [33]:
scores = cross_val_score(model,X_new,y_new,
                         cv=KFold(n_splits=4,shuffle=True,random_state=1234),
                         scoring='r2')

In [34]:
scores

array([0.13335827, 0.12922606, 0.13208168, 0.13031732])

In [35]:
scores.mean()

0.13124583285342406